app/src/main/java/ru/practicum/android/diploma/data/network/RetrofitNetworkClient.kt

```diff
@@ -0,0 +1,159 @@
+package ru.practicum.android.diploma.data.network
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.NetworkCapabilities
+import android.util.Log
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import retrofit2.HttpException
+import ru.practicum.android.diploma.data.dto.NetworkResponse
+import ru.practicum.android.diploma.data.dto.requests.AreaRequest
+import ru.practicum.android.diploma.data.dto.requests.CountriesRequest
+import ru.practicum.android.diploma.data.dto.requests.FindVacanciesRequest
+import ru.practicum.android.diploma.data.dto.requests.IndustriesRequest
+import ru.practicum.android.diploma.data.dto.requests.VacancyByIdRequest
+import ru.practicum.android.diploma.data.dto.responses.AreasListResponse
+import ru.practicum.android.diploma.data.dto.responses.CountriesListResponse
+import ru.practicum.android.diploma.data.dto.responses.IndustriesListResponse
+import java.io.IOException
+
+/**
+ * Error codes:
+ * -1 - IO EXCEPTION
+ * 100-500 - NETWORK ERROR CODES
+ */
+
+class RetrofitNetworkClient(
+    private val apiService: ApiService,
+    private val context: Context
+) : NetworkClient {
+
+    override suspend fun findVacancies(dto: FindVacanciesRequest): NetworkResponse {
+        if (!isConnected()) {
+            return NetworkResponse().apply { resultCode = ERROR_NO_INTERNET }
+        }
+
+        return withContext(Dispatchers.IO) {
+            try {
+                apiService.findVacancies(
+                    query = dto.query,
+                    salary = dto.salary,
+                    page = dto.page,
+                    amount = dto.amount,
+                    onlyWithSalary = dto.onlyWithSalary,
+                    industry = dto.industry,
+                    area = dto.area
+                ).apply { resultCode = SUCCESS_STATUS }
+            } catch (e: HttpException) {
+                Log.e(TAG, e.stackTraceToString())
+                NetworkResponse().apply { resultCode = e.code() }
+            } catch (e: IOException) {
+                Log.e(TAG, e.stackTraceToString())
+                NetworkResponse().apply { resultCode = IO_EXCEPTION }
+            }
+        }
+    }
+
+    override suspend fun getVacancyById(dto: VacancyByIdRequest): NetworkResponse {
+        if (!isConnected()) {
+            return NetworkResponse().apply { resultCode = ERROR_NO_INTERNET }
+        }
+
+        return withContext(Dispatchers.IO) {
+            try {
+                apiService.getVacancyById(dto.id).apply { resultCode = SUCCESS_STATUS }
+            } catch (e: HttpException) {
+                Log.e(TAG, e.stackTraceToString())
+                NetworkResponse().apply { resultCode = e.code() }
+            } catch (e: IOException) {
+                Log.e(TAG, e.stackTraceToString())
+                NetworkResponse().apply { resultCode = IO_EXCEPTION }
+            }
+        }
+    }
```

Comment:
Подобные конструкции предлагаю писать лаконичнее
```suggestion
    override suspend fun getVacancyById(dto: VacancyByIdRequest): NetworkResponse =
        if (!isConnected()) {
            NetworkResponse().apply { resultCode = ERROR_NO_INTERNET }
        } else {
             withContext(Dispatchers.IO) {
                try {
                    apiService.getVacancyById(dto.id).apply { resultCode = SUCCESS_STATUS }
                } catch (e: HttpException) {
                    Log.e(TAG, e.stackTraceToString())
                    NetworkResponse().apply { resultCode = e.code() }
                } catch (e: IOException) {
                    Log.e(TAG, e.stackTraceToString())
                    NetworkResponse().apply { resultCode = IO_EXCEPTION }
            }
        }
```